Add note on api compatibility (#862)
cert-manager/website: content/en/docs/installation/api_compatibility.md

---
title: "API compatibility"
linkTitle: "API compatibility"
weight: 70
type: "docs"

> ---

Reviewer: suggestion: I think everywhere we have a filename with multiple words in it, we use a dash rather than an underscore in the filename. e.g. [here](https://github.com/cert-manager/website/tree/ab263515a7c6e5e3280799710b42a072549a660d/content/en/docs/installation) we have `code-signing.md`, `kubectl-plugin.md`.

It doesn't _really_ matter of course, but I like that kind of consistency in the eventual URL.

cert-manager aims to abide by the same API compatibility policy as upstream Kubernetes APIs, see [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api).

> The main purpose of this is to ensure a smooth upgrade and downgrade experience for users, i.e to make sure that users' cert-manager custom resources keep functioning as before after an upgrade or downgrade of cert-manager.

Reviewer 2: ```suggestion
The main purpose of this is to ensure a smooth upgrade and downgrade experiences for users, i.e to make sure that users' cert-manager custom resources keep functioning as before after an upgrade or downgrade of cert-manager.
```

Reviewer: Is this suggestion not grammatically incorrect? I think this section is fine as-is.

Reviewer 2: Hmm, indeed, it's a half-step.
```suggestion
The main purpose of this is to ensure smooth upgrade and downgrade experiences for users, i.e to make sure that users' cert-manager custom resources keep functioning as before after an upgrade or downgrade of cert-manager.
```

The problem I have with the as-is content is that one can't both upgrade and downgrade concurrently, so it isn't one single experience. One can have other experiences.

> In some cases, we may need to require users to take actions before upgrade or may need to diverge from the API compatibility promise:

Reviewer 2: ```suggestion
In some cases, we may need to require users to take actions before upgrades or may need to diverge from the API compatibility promise:
```

(or `upgrading`)

Reviewer: +1 to "upgrading" here from me :+1:

- In cert-manager v1.7 [all alpha and beta API versions were removed](https://github.com/cert-manager/cert-manager/pull/4635). The Kubernetes Deprecation policy notes that API removal introduces an issue with objects stored at the removed versions. To fix this, we wrote a [custom tool](https://cert-manager.io/docs/installation/upgrading/remove-deprecated-apis/) that users could run once to migrate their resources.

> - In general, the main criteria of determining whether a change is acceptable should be user value. For example, in case of a critical bug, a fix that breaks the API compatibility promise by changing the default behavior of an API field _might_ be acceptable. (As of yet, there has not been a need for such a fix).

Reviewer: I like the wording of this a lot :+1:
